File: cpp/01_TwoSum.cpp
/*
 * Q1. Two Sum (Category: Arrays)
 *
 * Problem: Given an array of integers nums and an integer target,
 *          return indices of the two numbers such that they add up to target.
 *
 * Approach: Use an unordered_map to store element -> index.
 *           For each element, check if (target - element) exists in the map.
 *
 * Time Complexity : O(n)
 * Space Complexity: O(n)
 */
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> map; // value -> index
    for (int i = 0; i < (int)nums.size(); i++) {
        int complement = target - nums[i];
        if (map.count(complement)) {
            return {map[complement], i};
        }
        map[nums[i]] = i;
    }
    return {};
}

int main() {
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> result = twoSum(nums, target);
    cout << "Indices: [" << result[0] << ", " << result[1] << "]" << endl; // [0, 1]
    return 0;
}

File: python/01_two_sum.py
"""
Q1. Two Sum (Category: Arrays)

Problem: Given an array of integers nums and a target, return indices of the
         two numbers that add up to target.

Approach: Use a dict to map value -> index.
          For each element, check if (target - element) is already in the dict.

Time Complexity : O(n)
Space Complexity: O(n)
"""


def two_sum(nums: list[int], target: int) -> list[int]:
    seen = {}  # value -> index
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    return []


if __name__ == "__main__":
    print(two_sum([2, 7, 11, 15], 9))  # [0, 1]
    print(two_sum([3, 2, 4], 6))       # [1, 2]
